core/src/main/java/io/undertow/server/handlers/resource/CachingResourceManager.java

```diff
@@ -107,43 +112,61 @@ public CachedResource getResource(final String p) throws IOException {
         } else {
             path = p;
         }
-        Object res = cache.get(path);
-        if (res instanceof NoResourceMarker) {
-            NoResourceMarker marker = (NoResourceMarker) res;
-            long nextCheck = marker.getNextCheckTime();
-            if(nextCheck > 0) {
-                long time = System.currentTimeMillis();
-                if(time > nextCheck) {
-                    marker.setNextCheckTime(time + maxAge);
-                    if(underlyingResourceManager.getResource(path) != null) {
-                        cache.remove(path);
+        //ReadLock will allow multiple reads and changes, guarded by WriteLock for purge
+        final Lock readLock = this.cacheAccessLock.readLock();
+        try {
+            readLock.lock();
```

Comment:
getResource() method is incorrect I would say.
It should use both read lock & write lock separately (and not just read lock).

The first part under readlock should be:
---
Object res = null
readLock.lock();
try {
    res = cache.get(path);
} finally {
    readLock.unlock();
}

And later whenever an attempt to update cache field is made, write lock should be used instead.